[consistent-type-imports] False positive with decorator metadata

- [X] I have tried restarting my IDE and the issue persists.
- [X] I have updated to the latest version of the packages.
- [X] I have [read the FAQ](https://typescript-eslint.io/docs/linting/troubleshooting) and my problem is not listed.

**Repro**

```
(no rule configuration, using plugin:@typescript-eslint/all)
```

```TS
import type { Ref } from '@typegoose/typegoose';
import { prop } from '@typegoose/typegoose';

export class Test
{
    @prop({ ref: () => Test })
    public test?: Ref<Test>;
}
```

<!--
Also include your tsconfig, if you're using type-aware linting
-->

**Expected Result**

`Ref` is only used as a type, and is not used by the decorator, no issue should be raised here.

**Actual Result**

`ESLint: Type import "Ref" is used by decorator metadata.(@typescript-eslint/consistent-type-imports)`

However, typescript compiles the code fine with 'importsNotUsedAsValues'.

When removing `type`, typescript compilation fails:

```
../../src/backend/models/Test.ts:1:1 - error TS1371: This import is never used as a value and must use 'import type' 
because 'importsNotUsedAsValues' is set to 'error'.

1 import { Ref } from '@typegoose/typegoose';
  ~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
```

**Versions**

| package                            | version |
| ---------------------------------- | ------- |
| `@typescript-eslint/eslint-plugin` | `13.1.0` |
| `@typescript-eslint/parser`        | `5.12.1` |
| `TypeScript`                       | `4.5.5` |
| `ESLint`                           | `8.10.0` |
| `node`                             | `16.13.0` |

--- a/packages/eslint-plugin/tests/rules/consistent-type-imports.test.ts
+++ b/packages/eslint-plugin/tests/rules/consistent-type-imports.test.ts
@@ -14,6 +14,10 @@ const withMetaParserOptions = {
   project: './tsconfig-withmeta.json',
 };
 
+const withMetaConfigParserOptions = {
+  emitDecoratorMetadata: true,
+};
+
 ruleTester.run('consistent-type-imports', rule, {
   valid: [
     `
@@ -346,6 +350,113 @@ ruleTester.run('consistent-type-imports', rule, {
       `,
       parserOptions: withMetaParserOptions,
     },
+    {
+      code: `
+        import Foo from 'foo';
+        @deco
+        class A {
+          constructor(foo: Foo) {}
+        }
+      `,
+      parserOptions: withMetaConfigParserOptions,
+    },
+    {
+      code: `
+        import Foo from 'foo';
+        class A {
+          @deco
+          foo: Foo;
+        }
+      `,
+      parserOptions: withMetaConfigParserOptions,
+    },
+    {
+      code: `
+        import Foo from 'foo';
+        class A {
+          @deco
+          foo(foo: Foo) {}
+        }
+      `,
+      parserOptions: withMetaConfigParserOptions,
+    },
+    {
+      code: `
+        import Foo from 'foo';
+        class A {
+          @deco
+          foo(): Foo {}
+        }
+      `,
+      parserOptions: withMetaConfigParserOptions,
+    },
+    {
+      code: `
+        import Foo from 'foo';
+        class A {
+          foo(@deco foo: Foo) {}
+        }
+      `,
+      parserOptions: withMetaConfigParserOptions,
+    },
+    {
+      code: `
+        import Foo from 'foo';
+        class A {
+          @deco
+          set foo(value: Foo) {}
+        }
+      `,
+      parserOptions: withMetaConfigParserOptions,
+    },
+    {
+      code: `
+        import Foo from 'foo';
+        class A {
+          @deco
+          get foo() {}
+
+          set foo(value: Foo) {}
+        }
+      `,
+      parserOptions: withMetaConfigParserOptions,
+    },
+    {
+      code: `
+        import Foo from 'foo';
+        class A {
+          @deco
+          get foo() {}
+
+          set ['foo'](value: Foo) {}
+        }
+      `,
+      parserOptions: withMetaConfigParserOptions,
+    },
+    {
+      code: `
+        import type { Foo } from 'foo';
+        const key = 'k';
+        class A {
+          @deco
+          get [key]() {}
+
+          set [key](value: Foo) {}
+        }
+      `,
+      parserOptions: withMetaConfigParserOptions,
+    },
+    {
+      code: `
+        import * as foo from 'foo';
+        @deco
+        class A {
+          constructor(foo: foo.Foo) {}
+        }
+      `,
+      parserOptions: withMetaConfigParserOptions,
+    },
+
     // https://github.com/typescript-eslint/typescript-eslint/issues/2989
     `
 import type * as constants from './constants';
@@ -1521,6 +1632,166 @@ const a: Default = '';
         },
       ],
       parserOptions: withMetaParserOptions,
+    },
+    {
+      code: `
+        import type Foo from 'foo';
+        @deco
+        class A {
+          constructor(foo: Foo) {}
+        }
+      `,
+      output: noFormat`
+        import Foo from 'foo';
+        @deco
+        class A {
+          constructor(foo: Foo) {}
+        }
+      `,
+      errors: [
+        {
+          messageId: 'aImportInDecoMeta',
+          data: { typeImports: '"Foo"' },
+          line: 2,
+          column: 9,
+        },
+      ],
+      parserOptions: withMetaConfigParserOptions,
+    },
+    {
+      code: `
+        import type { Foo } from 'foo';
+        @deco
+        class A {
+          constructor(foo: Foo) {}
+        }
+      `,
+      output: noFormat`
+        import { Foo } from 'foo';
+        @deco
+        class A {
+          constructor(foo: Foo) {}
+        }
+      `,
+      errors: [
+        {
+          messageId: 'aImportInDecoMeta',
+          data: { typeImports: '"Foo"' },
+          line: 2,
+          column: 9,
+        },
+      ],
+      parserOptions: withMetaConfigParserOptions,
+    },
+    {
+      code: noFormat`
+        import type { Type } from 'foo';
+        import { Foo, Bar } from 'foo';
+        @deco
+        class A {
+          constructor(foo: Foo) {}
+        }
+        type T = Bar;
+      `,
+      output: noFormat`
+        import type { Type , Bar } from 'foo';
+        import { Foo } from 'foo';
+        @deco
+        class A {
+          constructor(foo: Foo) {}
+        }
+        type T = Bar;
+      `,
+      errors: [
+        {
+          messageId: 'aImportIsOnlyTypes',
+          data: { typeImports: '"Bar"' },
+          line: 3,
+          column: 9,
+        },
+      ],
+      parserOptions: withMetaConfigParserOptions,
+    },
+    {
+      code: `
+        import { V } from 'foo';
+        import type { Foo, Bar, T } from 'foo';
+        @deco
+        class A {
+          constructor(foo: Foo) {}
+          foo(@deco bar: Bar) {}
+        }
+      `,
+      output: noFormat`
+        import { V , Foo, Bar} from 'foo';
+        import type { T } from 'foo';
+        @deco
+        class A {
+          constructor(foo: Foo) {}
+          foo(@deco bar: Bar) {}
+        }
+      `,
+      errors: [
+        {
+          messageId: 'someImportsInDecoMeta',
+          data: { typeImports: '"Foo" and "Bar"' },
+          line: 3,
+          column: 9,
+        },
+      ],
+      parserOptions: withMetaConfigParserOptions,
+    },
+    {
+      code: `
+        import type { Foo, T } from 'foo';
+        import { V } from 'foo';
+        @deco
+        class A {
+          constructor(foo: Foo) {}
+        }
+      `,
+      output: noFormat`
+        import type { T } from 'foo';
+        import { V , Foo} from 'foo';
+        @deco
+        class A {
+          constructor(foo: Foo) {}
+        }
+      `,
+      errors: [
+        {
+          messageId: 'aImportInDecoMeta',
+          data: { typeImports: '"Foo"' },
+          line: 2,
+          column: 9,
+        },
+      ],
+      parserOptions: withMetaConfigParserOptions,
+    },
+    {
+      code: `
+        import type * as Type from 'foo';
+        @deco
+        class A {
+          constructor(foo: Type.Foo) {}
+        }
+      `,
+      output: noFormat`
+        import * as Type from 'foo';
+        @deco
+        class A {
+          constructor(foo: Type.Foo) {}
+        }
+      `,
+      errors: [
+        {
+          messageId: 'aImportInDecoMeta',
+          data: { typeImports: '"Type"' },
+          line: 2,
+          column: 9,
+        },
+      ],
+      parserOptions: withMetaConfigParserOptions,
     },
     {
       code: `

--- a/packages/parser/README.md
+++ b/packages/parser/README.md
@@ -67,6 +67,8 @@ interface ParserOptions {
 
   program?: import('typescript').Program;
   moduleResolver?: string;
+
+  emitDecoratorMetadata?: boolean;
 }
 ```
 
@@ -245,6 +247,12 @@ interface ModuleResolver {
 [Refer to the TypeScript Wiki for an example on how to write the `resolveModuleNames` function](https://github.com/Microsoft/TypeScript/wiki/Using-the-Compiler-API#customizing-module-resolution).
 
 Note that if you pass custom programs via `options.programs` this option will not have any effect over them (you can simply add the custom resolution on them directly).
+
+### `parserOptions.emitDecoratorMetadata`
+
+Default `undefined`.
+
+This option allow you to tell parser to act as if `emitDecoratorMetadata: true` is set in `tsconfig.json`, but without [type-aware linting](https://typescript-eslint.io/docs/linting/type-linting). In other words, you don't have to specify `parserOptions.project` in this case, making the linting process faster.
 
 ## Utilities
 

--- a/packages/parser/src/parser.ts
+++ b/packages/parser/src/parser.ts
@@ -132,6 +132,7 @@ function parseForESLint(
   const { ast, services } = parseAndGenerateServices(code, parserOptions);
   ast.sourceType = options.sourceType;
 
+  let emitDecoratorMetadata = options.emitDecoratorMetadata === true;
   if (services.hasFullTypeInformation) {
     // automatically apply the options configured for the program
     const compilerOptions = services.program.getCompilerOptions();
@@ -165,9 +166,12 @@ function parseForESLint(
       }
     }
     if (compilerOptions.emitDecoratorMetadata === true) {
-      analyzeOptions.emitDecoratorMetadata =
-        compilerOptions.emitDecoratorMetadata;
+      emitDecoratorMetadata = true;
     }
+  }
+
+  if (emitDecoratorMetadata) {
+    analyzeOptions.emitDecoratorMetadata = true;
   }
 
   const scopeManager = analyze(ast, analyzeOptions);

--- a/packages/types/src/parser-options.ts
+++ b/packages/types/src/parser-options.ts
@@ -37,6 +37,9 @@ interface ParserOptions {
   jsxFragmentName?: string | null;
   lib?: Lib[];
 
+  // use emitDecoratorMetadata without specifying parserOptions.project
+  emitDecoratorMetadata?: boolean;
+
   // typescript-estree specific
   comment?: boolean;
   debugLevel?: DebugLevel;

